File: android/app/src/main/kotlin/com/cruisemesh/app/mesh/CarriedOfferEpochGate.kt
package com.cruisemesh.app.mesh

import uniffi.cruisemesh_core.CoreCarriedOfferGate
import uniffi.cruisemesh_core.CoreCarriedOfferReservation
import uniffi.cruisemesh_core.coreCarriedOfferEpochMs

/**
 * Thin shell wrapper over the core's shared foreign-carry allowance (see
 * `CoreCarriedOfferGate` for the cap, the epoch length and the logical-peer
 * rule). iOS wraps the same core object, so neither the window nor the
 * allowance can drift between the platforms — the same arrangement as
 * [ReconnectBackoffTracker].
 *
 * Reservations are made before a digest plan is built. A non-empty plan is
 * [commit]ted and continues to count for the epoch; an empty or failed plan is
 * [release]d so it does not consume another peer's allowance.
 */
internal class CarriedOfferEpochGate(
    epochMs: Long = defaultEpochMs,
) {
    companion object {
        val defaultEpochMs: Long get() = coreCarriedOfferEpochMs()
    }

    private val core = CoreCarriedOfferGate.withEpochMs(epochMs)

    /**
     * The shared Rust gate itself, for `MessageStore.corePlanMeshMeet`, which
     * reserves and commits or releases the epoch slot inside one planning
     * call. It has to be this object, or the planner's encounters would spend
     * an allowance nobody else could see.
     */
    internal val coreState: CoreCarriedOfferGate get() = core

    val epochMs: Long get() = core.epochMs()

    /**
     * Claims one of the epoch's slots, or `null` when the allowance is spent or
     * this logical peer already had its offer. [logicalPeerId] is the peer's
     * UserID hex, never a link address.
     */
    fun tryReserve(nowMs: Long, logicalPeerId: String? = null): CoreCarriedOfferReservation? =
        core.tryReserve(nowMs, logicalPeerId)

    fun commit(reservation: CoreCarriedOfferReservation) = core.commit(reservation)

    fun release(reservation: CoreCarriedOfferReservation) = core.release(reservation)
}

File: ios/CruiseMesh/Mesh/CarriedOfferEpochGate.swift
import Foundation

/// Process-wide delegate to the core's shared foreign-carry allowance (Android
/// `CarriedOfferEpochGate.kt` twin).
///
/// There is deliberately no decision in this file: how many peers may offer
/// third-party traffic at once, how long the window lasts, and how one logical
/// peer reaching us at two addresses is collapsed to a single offer all live in
/// `core/src/transport_policy.rs`, so the two shells cannot drift.
///
/// Both lanes that offer foreign carry share this one allowance — the HELLO
/// drain and the digest spray — so a peer cannot be handed a full page by each
/// of them inside one connection burst.
///
/// Reservations are taken before a page is built. A page that comes out empty
/// is `release`d so it does not consume another peer's turn; one that actually
/// went out is `commit`ted and keeps counting until the window rolls.
///
/// This gates offering only. Nothing here removes a carried envelope or acks
/// anything: a deferred peer is offered on a later round, and a carried copy is
/// still retired only on digest-proof of receipt.
enum CarriedOfferEpochGate {
    private static let core = CoreCarriedOfferGate.withEpochMs(epochMs: coreCarriedOfferEpochMs())

    /// The shared Rust gate itself, for `MessageStore.corePlanMeshMeet`, which
    /// reserves and commits or releases the epoch slot inside one planning
    /// call. It has to be this object, or the planner's encounters would spend
    /// an allowance nobody else could see.
    static var coreState: CoreCarriedOfferGate { core }

    /// Claims one of the window's slots, or `nil` when the allowance is spent
    /// or this logical peer already had its offer. `logicalPeerId` is the
    /// peer's UserID hex, never a link address — keying it by address is what
    /// let one phone with two roles take every slot.
    static func tryReserve(
        nowMs: Int64,
        logicalPeerId: String? = nil
    ) -> CoreCarriedOfferReservation? {
        core.tryReserve(nowMs: nowMs, logicalPeerId: logicalPeerId)
    }

    static func commit(_ reservation: CoreCarriedOfferReservation) {
        core.commit(reservation: reservation)
    }

    static func release(_ reservation: CoreCarriedOfferReservation) {
        core.release(reservation: reservation)
    }
}
